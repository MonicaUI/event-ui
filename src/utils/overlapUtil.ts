/*
 * this function takes an array of date ranges in this format:
 * [{ startTime: ‘2018-07-03T19:53:58.570Z’, endTime: ‘2018-07-03T19:53:58.570Z’}]
 * the array is first sorted, and then checked for any overlap
 */

export function overlap(dateRanges: $TSFixMe): $TSFixMe {
  const sortedRanges = dateRanges.sort((previous, current) => {
    // get the start date from previous and current
    if (previous.startTime === undefined || current.startTime === undefined) {
      return 0;
    }
    const previousTime = new Date(previous.startTime).getTime();
    const currentTime = new Date(current.startTime).getTime();
    // if the previous is earlier than the current
    if (previousTime < currentTime) {
      return -1;
    }
    // if the previous time is the same as the current time
    if (previousTime === currentTime) {
      return 0;
    }
    // if the previous time is later than the current time
    return 1;
  });

  const overlapResult = sortedRanges.reduce(
    (result, current, idx, arr) => {
      // get the previous range
      if (idx === 0) {
        return result;
      }
      const previous = arr[idx - 1];
      // To match the service behavior, removing all null cases.
      if (
        previous.startTime === undefined ||
        current.endTime === undefined ||
        previous.endTime === undefined ||
        current.startTime === undefined
      ) {
        if (
          (current.endTime === undefined || previous.endTime === undefined) &&
          new Date(previous.startTime).getTime() === new Date(current.startTime).getTime()
        ) {
          result.overlap = true;
          result.ranges.push({
            previous,
            current
          });
        }
        return result;
      }
      // check for any overlap
      const previousEnd = new Date(previous.endTime).getTime();
      const currentStart = new Date(current.startTime).getTime();
      const overlapFlag = previous.id !== current.id && previousEnd > currentStart;

      // store the result
      if (overlapFlag) {
        // yes, there is overlap
        /* eslint no-param-reassign: ["error", { "props": true, "ignorePropertyModificationsFor": ["result"] }] */
        result.overlap = true;
        // store the specific ranges that overlap
        result.ranges.push({
          previous,
          current
        });
      }
      return result;
    },
    {
      overlap: false,
      ranges: []
    }
  );
  return overlapResult;
}
