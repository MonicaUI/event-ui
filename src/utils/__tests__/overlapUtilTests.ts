import { overlap } from '../overlapUtil';

const testData = [
  { id: 1, startTime: '2018-08-19T13:00:00.000Z', endTime: '2018-08-19T23:00:00.000Z' },
  { id: 2, startTime: '2018-07-19T13:00:00.000Z', endTime: '2018-07-19T23:00:00.000Z' },
  { id: 3, startTime: '2018-08-19T13:00:00.000Z', endTime: '2018-08-19T23:00:00.000Z' },
  { id: 4, startTime: '2018-07-19T13:00:00.000Z', endTime: '2018-07-19T23:00:00.000Z' },
  { id: 5, startTime: '2020-07-19T13:00:00.000Z', endTime: '2020-07-19T23:00:00.000Z' },
  { id: 6, startTime: '2019-07-19T13:00:00.000Z', endTime: '2019-07-19T23:00:00.000Z' }
];

test('overlapUtil matches snapshot', () => {
  expect(overlap(testData)).toMatchSnapshot();
});
