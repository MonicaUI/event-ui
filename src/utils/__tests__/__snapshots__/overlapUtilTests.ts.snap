// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`overlapUtil matches snapshot 1`] = `
Object {
  "overlap": true,
  "ranges": Array [
    Object {
      "current": Object {
        "endTime": "2018-07-19T23:00:00.000Z",
        "id": 4,
        "startTime": "2018-07-19T13:00:00.000Z",
      },
      "previous": Object {
        "endTime": "2018-07-19T23:00:00.000Z",
        "id": 2,
        "startTime": "2018-07-19T13:00:00.000Z",
      },
    },
    Object {
      "current": Object {
        "endTime": "2018-08-19T23:00:00.000Z",
        "id": 3,
        "startTime": "2018-08-19T13:00:00.000Z",
      },
      "previous": Object {
        "endTime": "2018-08-19T23:00:00.000Z",
        "id": 1,
        "startTime": "2018-08-19T13:00:00.000Z",
      },
    },
  ],
}
`;
